### PR TITLE
feat: filter markdown by query [5164]

### PR DESCRIPTION
**Pull request by Garzas**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [x] contains a reference JIRA issue number like `SQPIT-764`
    - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [ ] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

- **Structured Markdown Representation**: The application needed a more structured way to handle markdown content to improve readability and maintainability, and to ease UI rendering.
- **Markdown Content Searchability**: There was a need to filter and display markdown content based on user searches, showing only relevant parts of the documents.
- **Enhanced Display for Links and Lists**: Improvements were needed in how links and lists are displayed within markdown content, especially in the context of search results.

### Causes (Optional)

- The flat structure used for markdown representation made it hard to map content to UI components directly, complicating rendering and interactions.
- Users couldn't easily search within markdown documents, and even when possible, the context around search hits wasn't adequately presented.
- Lists and links within markdown didn't always render in a user-friendly manner, especially after applying filters or searches, leading to a disjointed user experience.

### Solutions

- **New Classes for Markdown**: Introduced a set of classes for a structured representation of markdown documents. This hierarchical approach mirrors the markdown's structure, facilitating direct mapping to UI components and improving code readability.
- **Filtering and Summarization**: Implemented advanced filtering for search queries within markdown content. The solution includes summarizing the content around search hits with ellipses for trimmed sections, ensuring users can understand the context of their search results.
- **UI Enhancements for Lists and Links**: Made specific adjustments for a more intuitive display of links and list items. This includes preserving the original order in lists, even in filtered views, and ensuring links are interactive and visually distinct. These changes aim to enhance the clarity and usability of markdown content in the UI.

### Attachments (Optional)

markdown_preview

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [3397](https://10.10.124.134/job/AR-build-pipeline/3397/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.7.0(9024164)(9024164)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.7.0(9024164)(9024164)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/ui/markdown/MarkdownComposer.kt [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /*
  * Wire
- * Copyright (C) 2024 Wire Swiss GmbH
+ * Copyright (C) 2023 Wire Swiss GmbH
```
Comment (saleniuk) [inline app/src/main/kotlin/com/wire/android/ui/markdown/MarkdownComposer.kt]: it's 2024 already, man 😄

File: app/src/main/kotlin/com/wire/android/ui/markdown/MarkdownComposer.kt [hunk 2]
```diff
@@ -318,11 +311,40 @@ fun appendLinksAndMentions(
...
+    }
+
+    return buildAnnotatedString {
+        append(text)
+        highlightIndexes
+            .forEach { highLightIndex ->
+                if (highLightIndex.endIndex <= length) {
+                    addStyle(
+                        style = SpanStyle(
+                            background = nodeData.colorScheme.highlight.copy(alpha = 0.5f),
```
Comment (saleniuk) [inline app/src/main/kotlin/com/wire/android/ui/markdown/MarkdownComposer.kt]: issue: all text highlight colors were updated to match designs in this PR: https://github.com/wireapp/wire-android/pull/2681
it should be:
```suggestion
                            background = nodeData.colorScheme.highlight,
                            color = nodeData.colorScheme.onHighlight,
```

**Comment by AndroidBob** [comment]
Build [3409](https://10.10.124.134/job/AR-build-pipeline/3409/) **failed**.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt [hunk 3]
```diff
@@ -895,14 +895,14 @@ fun MessageList(
...
     LaunchedEffect(lazyPagingMessages.itemCount) {
-        if (lazyPagingMessages.itemCount > prevItemCount.value) {
-            prevItemCount.value = lazyPagingMessages.itemCount
-
-            if (lazyListState.firstVisibleItemIndex > 0
+        if (lazyPagingMessages.itemCount > prevItemCount.value && selectedMessageId == null) {
+            if (prevItemCount.value > 0
+                && lazyListState.firstVisibleItemIndex > 0
                 && lazyListState.firstVisibleItemIndex <= MAXIMUM_SCROLLED_MESSAGES_UNTIL_AUTOSCROLL_STOPS
             ) {
```
Comment (yamilmedina) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt]:
```suggestion
           val canScrollToLastMessage = prevItemCount.value > 0
                && lazyListState.firstVisibleItemIndex > 0
                && lazyListState.firstVisibleItemIndex <= MAXIMUM_SCROLLED_MESSAGES_UNTIL_AUTOSCROLL_STOPS
            if (canScrollToLastMessage)
```

Suggestion: we can create a var to better express the condition here, like above 👆  or maybe a better name we can come up :D

**Comment by AndroidBob** [comment]
Build [3419](https://10.10.124.134/job/AR-build-pipeline/3419/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.7.0(9031544)(9031544)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.7.0(9031544)(9031544)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/ui/markdown/MarkdownComposer.kt [hunk 4]
```diff
@@ -29,180 +30,164 @@ import androidx.compose.ui.text.font.FontStyle
...
-        nodeData,
-        clickable
-    )
+    val filteredDocument = if (nodeData.searchQuery.isNotBlank()) {
+        document.filterNodesContainingQuery(nodeData.searchQuery)
+    } else {
+        document
+    }
+//    println("AFTER FILTER")
+//    printMarkdownNodeTree(filteredDocument)
```
Comment (yamilmedina) [inline app/src/main/kotlin/com/wire/android/ui/markdown/MarkdownComposer.kt]:
```suggestion

```

File: app/src/main/kotlin/com/wire/android/ui/markdown/MarkdownNode.kt [hunk 5]
```diff
@@ -0,0 +1,144 @@
+/*
```
Comment (yamilmedina) [inline app/src/main/kotlin/com/wire/android/ui/markdown/MarkdownNode.kt:1]: Loving it!

**Review by yamilmedina — COMMENTED** [review]
Just small comments and suggestions. Superb work, I like the expressiveness of the Markdown model now.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt (hunk 3 above)
Comment (Garzas) [reply app/src/main/kotlin/com/wire/android/ui/home/conversations/ConversationScreen.kt]: Good idea 👍

**Review by saleniuk — APPROVED** [review]

**Review by yamilmedina — APPROVED** [review]

**Comment by AndroidBob** [comment]
Build [3435](https://10.10.124.134/job/AR-build-pipeline/3435/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.7.0(9032750)(9032750)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.7.0(9032750)(9032750)-staging-compat.apk)

**Comment by AndroidBob** [comment]
Build [3494](https://10.10.124.134/job/AR-build-pipeline/3494/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.7.0(9083491)(9083491)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.7.0(9083491)(9083491)-staging-compat.apk)

**Pull request merged**